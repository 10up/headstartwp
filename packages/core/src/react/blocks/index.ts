export * from './ButtonBlock';
export * from './ButtonsBlock';
export * from './AudioBlock';
export * from './CodeBlock';
export * from './ColumnsBlocks';
export * from './ColumnBlock';
export * from './CoverBlock';
export * from './ParagraphBlock';
export * from './DebugBlock';
export * from './HeadingBlock';
export * from './QuoteBlock';
export * from './PullQuote';
export * from './PreformatedBlock';
export * from './VerseBlock';
export * from './TableBlock';
export * from './GroupBlock';
export * from './SeparatorBlock';
export * from './SpacerBlock';
export * from './ListBlock';
export * from './YoutubeLiteBlock';
