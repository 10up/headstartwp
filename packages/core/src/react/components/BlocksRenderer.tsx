import parse, { HTMLReactParserOptions, DOMNode, domToReact, Element } from 'html-react-parser';
import React, { isValidElement, FC, ReactNode, ReactElement } from 'react';
import type { IWhiteList } from 'xss';
import { wpKsesPost } from '../../dom';

export interface BlockProps {
	test: (domNome: DOMNode) => boolean;
	domNode?: DOMNode;
	children?: ReactNode | undefined;
}

interface BlockRendererProps {
	html: string;
	ksesAllowList?: IWhiteList;
	children?: ReactElement<BlockProps> | null;
}

export const BlocksRenderer: FC<BlockRendererProps> = ({ html, ksesAllowList, children }) => {
	const blocks: ReactNode[] = React.Children.toArray(children);

	// Check if components[] has a non-ReactNode type Element
	const hasInvalidComponent: boolean =
		blocks.findIndex(
			(block) => !isValidElement<BlockProps>(block) || !('test' in block.props),
		) !== -1;

	if (blocks.length > 1 && hasInvalidComponent) {
		console.warn(
			'Children of <BlocksRenderer /> component should be a type of ReactElement<BlockProps>',
		);
	}

	const cleanedHTML = wpKsesPost(html, ksesAllowList);

	const options: HTMLReactParserOptions = {
		replace: (domNode) => {
			let component: ReactNode = null;

			blocks.forEach((block) => {
				if (isValidElement<BlockProps>(block) && block.props.test(domNode)) {
					component = React.createElement(
						block.type,
						{
							...block.props,
							domNode,
						},
						domNode instanceof Element ? domToReact(domNode.children, options) : null,
					);
				}
			});

			return component;
		},
	};

	return <>{parse(cleanedHTML, options)}</>;
};

BlocksRenderer.defaultProps = {
	ksesAllowList: undefined,
	children: null,
};
