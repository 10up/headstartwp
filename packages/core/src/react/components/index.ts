export * from './BlocksRenderer';
export * from './BaseBlocksRenderer';
export * from './Menu';
export * from './SafeHtml';
export * from './HtmlDecoder';
