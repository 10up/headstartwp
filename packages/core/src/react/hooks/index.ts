export * from './useFetch';
export * from './useFetchPost';
export * from './useFetchPosts';
export * from './useFetchSearch';
export * from './types';
export * from './useFetchTerms';
