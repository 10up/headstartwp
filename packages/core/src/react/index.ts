export * from './components';
export * from './provider';
export * from './hooks';
export * from './blocks';
export * from './utils';
export * from '../dom';
