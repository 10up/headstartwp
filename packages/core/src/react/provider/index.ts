export * from './Provider';
export * from './types';
export * from './useSettings';
export * from './ThemeJsonProvider';
