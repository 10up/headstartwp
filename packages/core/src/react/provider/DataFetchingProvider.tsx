'use client';

import { FC } from 'react';
import { SWRConfig, SWRConfiguration } from 'swr';

export { unstable_serialize as serializeKey } from 'swr';

export type DataFetchingProviderProps = {
	/**
	 * Pass any configuration to the SWR library. Globally.
	 *
	 * These settings can be overridden at the hook level.
	 */
	swrConfig: SWRConfiguration;

	data: SWRConfiguration['fallback'];

	children: React.ReactNode;
};

export const DataFetchingProvider: FC<DataFetchingProviderProps> = ({
	swrConfig,
	data,
	children,
}) => {
	return <SWRConfig value={{ fallback: data, ...swrConfig }}>{children}</SWRConfig>;
};
