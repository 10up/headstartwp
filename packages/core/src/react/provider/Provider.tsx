import { PropsWithChildren, createContext, useMemo } from 'react';
import { getHeadlessConfig } from '../../utils/config';
import { SettingsContextProps } from './types';

export const SettingsContext = createContext<Partial<SettingsContextProps>>({});

interface ProviderProps {
	settings: SettingsContextProps;
}

export const SettingsProvider = ({ settings, children }: PropsWithChildren<ProviderProps>) => {
	const settingsValue = useMemo(
		() => ({
			...getHeadlessConfig(),
			...settings,
		}),
		[settings],
	);

	return <SettingsContext.Provider value={settingsValue}>{children}</SettingsContext.Provider>;
};
