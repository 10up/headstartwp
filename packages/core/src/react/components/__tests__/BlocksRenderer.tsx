import { render } from '@testing-library/react';
import { DOMNode, domToReact, Element } from 'html-react-parser';
import React, { ReactElement } from 'react';
import { isAnchorTag } from '../../../dom';
import { SettingsProvider } from '../../provider';
import { BlocksRenderer } from '../BlocksRenderer';
import type { BlockProps } from '../BaseBlocksRenderer';

describe('BlocksRenderer', () => {
	it('renders html properly', () => {
		const { container } = render(<BlocksRenderer html="<div><p>hello world</p></div>" />);
		expect(container.firstChild).toMatchInlineSnapshot(`
      <div>
        <p>
          hello world
        </p>
      </div>
    `);
	});

	it('sanitizes html', () => {
		const { container } = render(
			<BlocksRenderer html="<div><p><script>alert()</script>hello world</p></div>" />,
		);
		expect(container.firstChild).toMatchInlineSnapshot(`
      <div>
        <p>
          hello world
        </p>
      </div>
    `);
	});

	it('sanitizes html respecting ksesAllowList', () => {
		const { container } = render(
			<BlocksRenderer
				html="<div><p>hello world</p> div content</div>"
				ksesAllowList={{ div: [] }}
			/>,
		);
		expect(container.firstChild).toMatchInlineSnapshot(`
      <div>
         div content
      </div>
    `);
	});

	it('replaces markup with react components', () => {
		const DivToP = ({ domNode, children }: BlockProps) => {
			const className =
				domNode instanceof Element ? domNode?.attribs.class || undefined : undefined;
			return <p className={className}>{children}</p>;
		};

		const { container } = render(
			<BlocksRenderer html="<div class='my-class'>This Will Become a p tag</div><div>This Will Become a p tag</div>">
				<DivToP
					test={(node: DOMNode) => {
						if (!(node instanceof Element)) {
							return false;
						}
						return node.type === 'tag' && node.name === 'div';
					}}
				/>
			</BlocksRenderer>,
		);

		expect(container).toMatchInlineSnapshot(`
      <div>
        <p
          class="my-class"
        >
          This Will Become a p tag
        </p>
        <p>
          This Will Become a p tag
        </p>
      </div>
    `);
	});

	it('replaces markup with react components without test function', () => {
		const DivToP = ({ domNode, children }: BlockProps) => {
			const className =
				domNode instanceof Element ? domNode?.attribs.class || undefined : undefined;
			return <p className={className}>{children}</p>;
		};

		const { container } = render(
			<BlocksRenderer html="<div class='my-class'>This Will Become a p tag</div><div>This Will not Become a p tag</div>">
				<DivToP tagName="div" classList={['my-class']} />
			</BlocksRenderer>,
		);

		expect(container).toMatchInlineSnapshot(`
      <div>
        <p
          class="my-class"
        >
          This Will Become a p tag
        </p>
        <div>
          This Will not Become a p tag
        </div>
      </div>
    `);
	});

	it('replaces markup with react components and remove children nodes', () => {
		const DivToP = ({ domNode, children }: BlockProps) => {
			const className =
				domNode instanceof Element ? domNode?.attribs.class || undefined : undefined;
			return <p className={className}>{children}</p>;
		};

		const { container } = render(
			<BlocksRenderer html="<div class='my-class'>This Will Become a p tag<pre>this will be removed</pre></div><div>This Will not Become a p tag</div>">
				<DivToP
					tagName="div"
					classList={['my-class']}
					exclude={(node) => node.name === 'pre'}
				/>
			</BlocksRenderer>,
		);

		expect(container).toMatchInlineSnapshot(`
      <div>
        <p
          class="my-class"
        >
          This Will Become a p tag
        </p>
        <div>
          This Will not Become a p tag
        </div>
      </div>
    `);
	});

	it('passes the current site to test function', () => {
		const InternalLink: React.FC<BlockProps> = () => {
			return <p>INTERNAL LINK</p>;
		};

		const { container } = render(
			<SettingsProvider settings={{ sourceUrl: 'https://sourceurl.com/site1' }}>
				<BlocksRenderer html="<a href='https://sourceurl.com/site1/post-name'>link</a><a href='https://sourceurl.com/site2/post-name'>not internal link</a>">
					<InternalLink
						test={(node, site) => {
							return isAnchorTag(node, { isInternalLink: true }, site);
						}}
					/>
				</BlocksRenderer>
			</SettingsProvider>,
		);

		expect(container).toMatchInlineSnapshot(`
      <div>
        <p>
          INTERNAL LINK
        </p>
        <a
          href="https://sourceurl.com/site2/post-name"
        >
          not internal link
        </a>
      </div>
    `);
	});

	it('passes the inline styles as an object to the component', () => {
		const FlexDiv: React.FC<BlockProps> = ({ domNode, style, children }) => {
			if (style && style.display !== 'flex') {
				style.display = 'flex';

				return <div style={style}>{children}</div>;
			}

			return domToReact([domNode as DOMNode]) as ReactElement;
		};

		const { container } = render(
			<BlocksRenderer html="<div style='display:flex;'>first content</div><div style='display:block;'>second content</div>">
				<FlexDiv
					test={(node) => {
						return node.name === 'div';
					}}
				/>
			</BlocksRenderer>,
		);

		expect(container).toMatchInlineSnapshot(`
      <div>
        <div
          style="display: flex;"
        >
          first content
        </div>
        <div
          style="display: flex;"
        >
          second content
        </div>
      </div>
	`);
	});
});
