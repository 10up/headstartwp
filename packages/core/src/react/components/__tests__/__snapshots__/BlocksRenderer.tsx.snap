// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlocksRenderer forward blockProps to the component 1`] = `
<div>
  <p
    class="my-class"
  >
    {"attributes":{"blockAttribute":"this is a block attribute"},"name":"10up/custom-block","className":"my-class"}
  </p>
</div>
`;

exports[`BlocksRenderer forward context to the component 1`] = `
<div>
  <p
    class="my-class"
  >
    {"attributes":{"blockAttribute":"this is a block attribute"},"name":"10up/custom-block","className":"my-class"}
     - 
    {"contextProp":"this is a context prop"}
  </p>
</div>
`;

exports[`BlocksRenderer works correctly with chinese content 1`] = `
<div>
  <h3
    class="wp-block-heading"
    data-wp-block="{"level":3,"hash":"9fb25fc5-7456-4704-bffe-aa438487253b"}"
    data-wp-block-name="core/heading"
  >
    <strong>
      æ´»å‹•æœŸé–“
    </strong>
  </h3>
</div>
`;
